File: Python/PE019.py
"""
Problem 19: Counting Sundays
URL: https://projecteuler.net/problem=19

    You are given the following information, but you may prefer
    to do some research for yourself.

        + 1 Jan 1900 was a Monday.
        + Thirty days has September,
          April, June and November.
          All the rest have thirty-one,
          Saving February alone,
          Which has twenty-eight, rain or shine.
          And on leap years, twenty-nine.
        + A leap year occurs on any year evenly divisible by 4,
          but not on a century unless it is divisible by 400.

    How many Sundays fell on the first of the month during the
    twentieth century (1 Jan 1901 to 31 Dec 2000)?
"""
import datetime
from shared import check_argv

######################################################################
# Constants
######################################################################

SUNDAY = 6
DATE_BEGIN = "1901-01-01"
DATE_END = "2000-12-31"


######################################################################
# Functions
######################################################################

def add_month(value):
    """
    Adds one month to a date.
    :param value: The date to change.
    :return: The new date.
    """
    if value.month < 12:
        year = value.year
        month = value.month + 1
        day = value.day
    else:
        year = value.year + 1
        month = 1
        day = value.day
    return datetime.date(year, month, day)


def get_weekdays(begin, end, weekday):
    """
    Gets a list of dates of those 1st of the months that
    fell on a given weekday inside a range of dates.
    :param begin: The begin date.
    :param end: The end date.
    :param weekday: The week day to select.
    :return: The list of days that match with the weekday.
    """
    dates = []
    current = datetime.date.fromisoformat(begin)
    limit = datetime.date.fromisoformat(end)
    while current <= limit:
        if current.weekday() == weekday:
            dates.append(current.strftime("%A, %d-%B-%Y"))
        current = add_month(current)
    return dates


def main():
    """
    Main entry for the problem solver.
    """
    dates = get_weekdays(DATE_BEGIN, DATE_END, SUNDAY)
    result = len(dates)

    # Show the final dates of the problem:
    if check_argv("show_dates"):
        print("List of obtained dates:")
        for date in dates:
            print(date)
        print()

    # Show the final result of the problem:
    print(f"The number of Sundays that fell on the month's 1st during the 20th century is {result}.")


######################################################################
# Module execution
######################################################################

if __name__ == "__main__":
    main()

File: Erlang/src/pe019.erl
%%%======================================================================
%%% @author Gorka Suárez García
%%% @copyright (C) 2022-2023, Gorka Suárez García
%%% @doc
%%% You are given the following information, but you may prefer
%%% to do some research for yourself.
%%%
%%%     + 1 Jan 1900 was a Monday.
%%%     + Thirty days has September,
%%%       April, June and November.
%%%       All the rest have thirty-one,
%%%       Saving February alone,
%%%       Which has twenty-eight, rain or shine.
%%%       And on leap years, twenty-nine.
%%%     + A leap year occurs on any year evenly divisible by 4,
%%%       but not on a century unless it is divisible by 400.
%%%
%%% How many Sundays fell on the first of the month during the
%%% twentieth century (1 Jan 1901 to 31 Dec 2000)?
%%% @end
%%%======================================================================
-module(pe019).
-author("Gorka Suárez García").
-export([main/0, result/0]).

-define(SUNDAY, 7).
-define(DATE_BEGIN, {1901, 1, 1}).
-define(DATE_LIMIT, {2001, 1, 1}).

%%-----------------------------------------------------------------------
%% @doc
%% Main entry for the problem solver.
%% @end
%%-----------------------------------------------------------------------
main() ->
    io:format("The number of Sundays that fell on the month's 1st during the 20th century is ~p.~n", [result()]).

%%-----------------------------------------------------------------------
%% @doc
%% Main result for the problem solver.
%% @end
%%-----------------------------------------------------------------------
result() ->
    Dates = get_weekdays(?DATE_BEGIN, ?DATE_LIMIT, ?SUNDAY),
    length(Dates).

%%-----------------------------------------------------------------------
%% @private
%% @doc
%% Gets a list of dates of those 1st of the months that
%% fell on a given weekday inside a range of dates.
%% @param Begin The begin date.
%% @param Limit The limit date.
%% @param Weekday The week day to select.
%% @returns The list of days that match with the weekday.
%% @end
%%-----------------------------------------------------------------------
get_weekdays(Begin, Limit, Weekday) ->
    get_weekdays(Begin, Limit, Weekday, []).

%%-----------------------------------------------------------------------
%% @private
%% @doc
%% Internal function for 'fun get_weekdays/3'.
%% @end
%%-----------------------------------------------------------------------
get_weekdays(Current = {Year, Month, Day}, Limit, Weekday, Accumulated) ->
    case Current < Limit of
        false ->
            lists:reverse(Accumulated);
        true ->
            Next = case calendar:day_of_the_week(Year, Month, Day) of
                       Weekday -> [Current|Accumulated];
                       _ -> Accumulated
                   end,
            get_weekdays(add_month(Current), Limit, Weekday, Next)
    end.

%%-----------------------------------------------------------------------
%% @private
%% @doc
%% Adds one month to a date.
%% @param Value The date to change.
%% @returns The new date.
%% @end
%%-----------------------------------------------------------------------
add_month({Year, Month, Day}) ->
    case Month < 12 of
        false -> {Year + 1, 1, Day};
        true -> {Year, Month + 1, Day}
    end.
